### PR TITLE
Track vector dimensions in inverted index

### PR DESCRIPTION
**Pull request by parkerduckworth**
### Review checklist

- [ ] Documentation has been updated, if necessary. Link to changed documentation:
- [ ] Chaos pipeline run or not necessary. Link to pipeline:
- [ ] All new code is covered by tests where it is reasonable.
- [ ] Performance tests have been run or not necessary.

### CODE REVIEW

File: adapters/repos/db/shard_dimension_tracking.go [hunk 1]
```diff
@@ -0,0 +1,51 @@
...
+	return sum
+}
+
+func (s *Shard) initDimensionTracking() {
+	if !s.index.Config.TrackVectorDimensions {
+		return
+	}
+
+	go func() {
+		t := time.NewTicker(5 * time.Second) // 5 minutes
```
Comment (antas-marcin) [inline adapters/repos/db/shard_dimension_tracking.go:33]: Comment says 5 minutes but ticker is set to 5 seconds, which is true?

File: adapters/repos/db/vector/noop/index.go [hunk 2]
```diff
@@ -75,3 +75,6 @@ func (i *Index) ListFiles(context.Context) ([]string, error) {
 func (i *Index) ResumeMaintenance(context.Context) error {
 	return nil
 }
+
+func (i *Index) Dump(labels ...string) {
```
Comment (antas-marcin) [inline adapters/repos/db/vector/noop/index.go:79]: why this method is empty?

File: adapters/repos/db/vector_index.go [hunk 3]
```diff
@@ -21,6 +21,7 @@ import (
 // VectorIndex is anything that indexes vectors efficiently. For an example
 // look at ./vector/hsnw/index.go
 type VectorIndex interface {
+	Dump(labels ...string)
```
Comment (antas-marcin) [inline adapters/repos/db/vector_index.go:24]: Is this method implemented?

File: usecases/monitoring/prometheus.go [hunk 4]
```diff
@@ -230,6 +231,10 @@ func newPrometheusMetrics() *PrometheusMetrics {
 			Name: "backup_store_data_transferred",
 			Help: "Total number of bytes transferred during a backup store",
 		}, []string{"backend_name", "class_name"}),
+		DimensionSum: promauto.NewGaugeVec(prometheus.GaugeOpts{
+			Name: "dimension_sum",
+			Help: "Total dimensions in a shard",
+		}, []string{"class_name", "shard_name"}),
```
Comment (antas-marcin) [inline usecases/monitoring/prometheus.go:237]: one comment on the naming, wouldn't it be better to find if the metric would be called: `vector_dimensions_sum`?

File: adapters/repos/db/shard_dimension_tracking.go (hunk 1 above)
Comment (etiennedi) [reply adapters/repos/db/shard_dimension_tracking.go:33]: I think this comment was originally from me. The motivation was that it happened super frequently during the development (5s), but once it's done, we can have it happen less often (5min) as there is some cost to this. (The cost is rather low due to the inverted index approach, but it's not free).

File: adapters/repos/db/vector_index.go (hunk 3 above)
Comment (donomii) [reply adapters/repos/db/vector_index.go:24]: Visual studio isn't showing the implementations, but it is defined.  I don't think go will compile successfully unless all the functions are implemented.

File: usecases/monitoring/prometheus.go [hunk 5]
```diff
@@ -230,6 +231,10 @@ func newPrometheusMetrics() *PrometheusMetrics {
 			Name: "backup_store_data_transferred",
 			Help: "Total number of bytes transferred during a backup store",
 		}, []string{"backend_name", "class_name"}),
+		DimensionSum: promauto.NewGaugeVec(prometheus.GaugeOpts{
+			Name: "vector_dimension_sum",
```
Comment (antas-marcin) [inline usecases/monitoring/prometheus.go:235]: shouldn't be `vector_dimensions_sum` instead of `vector_dimension_sum`?

**Pull request merged**